vault/vault.go

```diff
@@ -0,0 +1,128 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package vault provides helper functions to improve the go-metrics to stackdriver metric
+// conversions specific to HashiCorp Vault.
+package vault
+
+import "github.com/armon/go-metrics"
+
+// Extractor extracts known patterns from the key into metrics.Label for better metric grouping
+// and to help avoid the limit of 500 custom metric descriptors per project
+// (https://cloud.google.com/monitoring/quotas).
+func Extractor(key []string) ([]string, []metrics.Label) {
+	if len(key) == 2 {
+		if key[0] == "database" {
+			return key[:1], []metrics.Label{
+				{
+					Name:  "method",
+					Value: key[1],
+				},
+			}
+		}
+	}
+	if len(key) == 3 {
+		if key[0] == "vault" && (key[1] == "barrier" || key[1] == "token" || key[1] == "policy") {
```

Comment:
I think a small helper could help make this code a bit more readable:

```go
func oneOf(k string, p ...string) bool {
  for _, v := range p {
    if k == v {
      return true
    }
  }
  return false
}
```

Then this becomes:

```go
if key[0] == "vault" && oneOf(key[1], "barrier", "token", "policy") { // ...
```

Maybe it's not worth it /shrug